libvast/src/system/type_registry.cpp

```diff
@@ -88,11 +88,48 @@ caf::error type_registry_state::load_from_disk() {
 }
 
 void type_registry_state::insert(vast::type layout) {
-  auto& cont = data[layout.name()].value;
+  auto flattened_layout = flatten(layout);
+  auto& old_layouts = data[flattened_layout.name()].value;
+  // Helper function.
+  auto is_compatible = [](const type& x) noexcept {
+    auto lhs = caf::get_if<record_type>(&x);
+    return [&, lhs](const type& y) noexcept {
+      auto rhs = caf::get_if<record_type>(&y);
+      // If they are not record types, check if they are congruent.
+      if (!lhs || !rhs)
+        return congruent(x, y);
+      // Check whether all fields of rhs exist in lhs, i.e., no new fields were
+      // added.
+      for (auto&& rf : rhs->fields) {
+        if (auto lf = lhs->find(rf.name)) {
+          // If there's a type mismatch, exit early as well.
+          if (!congruent(lf->type, rf.type))
+            return false;
+        } else {
+          // Not all fields of rhs exist in lhs. Exit early.
+          return false;
+        }
+      }
+      // We passed the test for all fields. Yay!
+      return true;
+    };
+  };
```

Comment:
Would anything prevent against beefing up the function `congruent` with that logic? Then it could also be truly recursive.

Reply:
I'd rather duplicate the logic and call it `compatible`. This is not about congruency. Note that as is this expects flattened types.